### ISSUE TITLE
Add support for Ruby 3.0.0

### ISSUE THREAD

**Issue by adfoster-r7** [rapid7/metasploit-framework#14666]
## Summary

Update Metasploit framework to support Ruby 3.0.0

### Motivation

Ruby 2.x will eventually EOL and stop receiving security upgrades
Kali etc will most likely ship with Ruby 3.0.0, which Metasploit currently does not work with

#### 1. Updating pcaprub

pcaprub does not currently install as a dependency, related pull request: https://github.com/pcaprub/pcaprub/pull/58

```
Fetching gem metadata from https://rubygems.org/.........
pcaprub-0.13.0 requires ruby version ~> 2.0, which is incompatible with the current version, ruby 3.0.0p0
```

Current work around within Gemfile:
```
# Note: Additionally requires a `bundle` in framework, and compiling pcaprub manually: `cd $(bundle info --path pcaprub) && rake compile`
gem 'pcaprub', '0.13.1',
    git: 'https://github.com/pcaprub/pcaprub',
    branch: 'ruby3',
    require: false
```

And for the eventmachine compilation issue on mac:
```
gem install eventmachine -- --with-openssl-dir=/usr/local/opt/openssl@1.1
```

#### 2. Update rex-core to explicitly depend on webrick

Metasploit data models has an implicit dependency on webrick [here](https://github.com/rapid7/metasploit_data_models/blob/ac04af7d3440711e85ceb9f68e51c6d4a91d437c/lib/metasploit_data_models.rb#L5-L6)
 which breaks and must be updated to work:

> +#/Users/user/.rvm/gems/ruby-3.0.0@metasploit-framework/gems/metasploit_data_models-4.1.1/lib/metasploit_data_models.rb:6:in `require': cannot load such file -- webrick (LoadError)

Webrick is no longer included in Ruby 3.0.0 by default:
https://www.ruby-lang.org/en/news/2020/12/25/ruby-3-0-0-released/

> The following libraries are no longer bundled gems or standard libraries. Install the corresponding gems to use these features.
> - sdbm
> - webrick
> - net-telnet
> - xmlrpc

####  3. Upgrade to Rails 6

Ruby 3.0.0 stops Metasploit console from starting up due to currently being on Rails 5.2

#### 4. Update uses of `rexml`

Continuation of point 1, the rex-exploitation gem uses `rexml`

#### Fix Ruby deprecation issues

1. SQLite3 deprecation warning

> /usr/share/metasploit-framework/modules/post/windows/gather/enum_chrome.rb:144: warning: rb_check_safe_obj will be removed in Ruby 3.0

Looks like this is fixed by in sqlite3-ruby 1.4.2:

https://github.com/sparklemotion/sqlite3-ruby/blame/d94c42b0b8e5005ee28cde75a58e15ea5670d1c4/CHANGELOG.rdoc#L1-L6

However we're currently pinned to an older version: https://github.com/rapid7/metasploit-framework/pull/11363

2. Remove tainted calls from rex-core

https://github.com/rapid7/rex-core/pull/11

**Comment by kuzuri-03** [comment]
Can anyone give the solution to this problem?

root@kali:/usr/share/metasploit-framework# msfconsole
Could not find aws-partitions-1.417.0 in any of the sources
Run `bundle install` to install missing gems.
root@kali:/usr/share/metasploit-framework#  bundle install
Don't run Bundler as root. Bundler can ask for sudo if it is needed, and
installing your bundle as root will break this application for all non-root
users on this machine.
Fetching gem metadata from https://rubygems.org/............
pcaprub-0.13.0 requires ruby version ~> 2.0, which is incompatible with the
current version, ruby 3.0.0p0

There is no proper guidance to tackle this step !
please help @elia @adfoster-r7 @trevrosen @techpeace @badboy

**Comment by badboy** [comment]
No idea why you tagged me (or a couple of other people). That's kinda rude. If you would have spent just a tiny bit of time on this you should have seen that I am most likely not the right person to help you here.

Demanding a solution on a problem that has been filed a mere day ago is also kinda rude.
It would be best if you give the maintainers some time to react. They might be busy with a ton of other stuff as well.

The original bug report has some steps that are required to get this to work. Why don't you start helping out there?

**Comment by techpeace** [comment]
The guidance is in the command's first line of output:

"Don't run Bundler as root."

**Comment by adfoster-r7** [comment]
@sbrun Would you happen to know when Kali plans to use Ruby 3.0.0? We would like to aim for Metasploit being fully updated before that date, so any insight would be appreciated :+1:

**Comment by sbrun** [comment]
@adfoster-r7 Kali follows Debian Testing and Debian plans to use Ruby 3.0.0 after the Bullseye release (the freeze started). No date is set, Debian release is done when "it's ready". I think it will be around this summer.

**Comment by capunk2708** [comment]
I also experienced the same thing
![Uploading Screenshot_20210211_071411.jpg…]()

**Comment by anthraxx** [comment]
The time has come, Ruby 3.0 is arriving in Arch Linux, which now hits this issue unfortunately. :cat2:

**Comment by Mayor294** [comment]
Pls try upgrading Metasploit to work with ruby 3 plss...

Reasons this is what happens when it's making use of old version of Ruby 

CANNOT LINK EXECUTABLE "ruby": library "/data/data/com.termux/files/usr/lib/ruby/2.7.0/aarch64-linux-android/bigdecimal.so" not found 

And besides I'm stuck I don't know what next to do I've tried everything possible yet it still show that 😒😒...

Had it been it's ruby 3.0 this wouldn't have happened

**Comment by armalitedev** [comment]
> Pls try upgrading Metasploit to work with ruby 3 plss...
> 
> Reasons this is what happens when it's making use of old version of Ruby
> 
> CANNOT LINK EXECUTABLE "ruby": library "/data/data/com.termux/files/usr/lib/ruby/2.7.0/aarch64-linux-android/bigdecimal.so" not found
> 
> And besides I'm stuck I don't know what next to do I've tried everything possible yet it still show that 😒😒...
> 
> Had it been it's ruby 3.0 this wouldn't have happened

same bro

**Comment by ilovesusu** [comment]
2021-08-14 debian 11 Release, Ruby3.0 about to happen

**Comment by adfoster-r7** [comment]
The final pull request for Ruby 3 support is here https://github.com/rapid7/metasploit-framework/pull/15537, hoping to get that out for next week's framework release. Note that there's a bug in Ruby itself which will require at least Ruby 3.0.1 to work with framework.

**Comment by jeffmcjunkin** [comment]
Just checking, is Ruby 3.0.1 now supported in the nightly builds? I'm getting caught on #14763 again, and `insert_all` API sounds like it'd result in some big speed increases, which depends on Ruby 3.0.

FWIW I'm seeing slowdowns (and imports stopping entirely) on moderately-sized masscan XML imports (~5000 hosts) as well. A quick profile of postgres shows the excessive `SELECT` statements.

**Comment by adfoster-r7** [comment]
@jeffmcjunkin The [insert_all](https://apidock.com/rails/v6.0.0/ActiveRecord/Persistence/ClassMethods/insert_all) API is available now as part of our [Rails 6.1 upgrade](https://github.com/rapid7/metasploit-framework/pull/15493), which was a separate effort to the [Ruby 3.x support](https://github.com/rapid7/metasploit-framework/pull/15537). We should revisit the slow `db_import` functionality again, thanks for the heads up :+1:

**Comment by adfoster-r7** [comment]
Closing this issue as Metasploit should now support Ruby 3.0.1
